File: cpp/p3.cpp
#include <cmath>
#include <cstdlib>
#include <iostream>

int main(int argc, char *argv[]) {
  unsigned long int n = 600851475143;
  if (argc > 1) {
    n = atoi(argv[1]);
  }
  unsigned long int last_factor = 1;
  if (n % 2 == 0) {
    last_factor = 2;
    n /= 2;
    while (n % 2 == 0) {
      n /= 2;
    }
  }
  unsigned long int factor = 3;
  unsigned long int max_factor = sqrt(n);
  while (n > 1 && factor <= max_factor) {
    if (n % factor == 0) {
      n /= factor;
      last_factor = factor;
      while (n % factor == 0) {
        n /= factor;
      }
      max_factor = sqrt(n);
    }
    factor += 2;
  }
  if (n != 1) {
    last_factor = n;
  }
  std::cout << last_factor << std::endl;
}

File: fsharp/p3.fs
[<EntryPoint>]
let main(argv) =
  let mutable n:int64 = if Array.length argv > 0 then argv.[0] |> int64
                        else 600851475143L
  let mutable last_factor:int64 = 1L
  if n % 2L = 0L then
    last_factor <- 2L
    n <- n / 2L
    while n % 2L = 0L do
      n <- n / 2L
  let mutable factor:int64 = 3L
  let mutable max_factor:int64 = sqrt (n |> float) |> int64
  while n > 1L && factor <= max_factor do
    if n % factor = 0L then
      n <- n / factor
      last_factor <- factor
      while n % factor = 0L do
        n <- n / factor
      max_factor <- sqrt (n |> float) |> int64
    factor <- factor + 2L
  if n <> 1L then
    last_factor <- n
  printfn "%i" last_factor
  0
